presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -377,19 +398,113 @@ public TableStatistics getTableStatistics(ConnectorTransactionHandle transaction
+                        case STRING:
+                            return OptionalDouble.of(columnStatistics.getStringColumnStatistics().getNullsCount());
+                        case BINARY:
+                            return OptionalDouble.of(columnStatistics.getBinaryColumnStatistics().getNullsCount());
+                        default:
+                            return OptionalDouble.empty();
+                    }
+                },
+                doubleStream -> {
+                    List<Double> values = doubleStream.mapToObj(Double::valueOf).collect(toList());
+                    long valuesCount = values.size();
+                    double valuesSum = values.stream().mapToDouble(Double::doubleValue).sum();
+                    if (valuesCount == 0) {
+                        return OptionalDouble.empty();
+                    }
+                    else {
+                        return OptionalDouble.of(partitionStatisticsMap.size() / valuesCount * valuesSum);
+                    }
+                });
+    }
+
+    private Estimate getColumnStatisticsValue(
+            Map<String, PartitionStatistics> partitionStatisticsMap,
+            String column,
+            Function<ColumnStatistics, OptionalDouble> valueExtractFunction,
+            Function<DoubleStream, OptionalDouble> valueAggregateFunction)
+    {
+        DoubleStream intermediateStream = partitionStatisticsMap.values().stream()
+                .map(PartitionStatistics::getColumnStatistics)
+                .filter(columnStatisticsMap -> columnStatisticsMap.containsKey(column))
+                .map(columnStatisticsMap -> columnStatisticsMap.get(column))
+                .map(valueExtractFunction)
+                .filter(OptionalDouble::isPresent)
+                .mapToDouble(OptionalDouble::getAsDouble);
+
+        OptionalDouble statisicsValue = valueAggregateFunction.apply(intermediateStream);
+
+        if (statisicsValue.isPresent()) {
+            return new Estimate(statisicsValue.getAsDouble());
+        }
+        else {
+            return Estimate.unknownValue();
+        }
+    }
 
-        TableStatistics tableStatistics = EMPTY_STATISTICS;
-        if (partitionsWithStatsCount > 0) {
-            tableStatistics = TableStatistics.builder()
-                    .setRowCount(new Estimate(1.0 * numRowsSum / partitionsWithStatsCount * allPartitionsCount))
-                    .build();
-            // TODO uncertainity level should be handled here
+    private Map<String, ColumnHandle> getTableColumnNames(ConnectorTableLayoutHandle layoutHandle, ConnectorSession session)
+    {
+        HiveTableLayoutHandle layout = checkType(layoutHandle, HiveTableLayoutHandle.class, "layoutHandle");
+        if (!layout.getPartitions().isPresent()) {
+            return ImmutableMap.of();
         }
-        return tableStatistics;
+        HivePartition hivePartition = Iterables.getFirst(layout.getPartitions().get(), null);
```

Comment:
use `stream(layout).findAny(layout -> layout.equals(...))`. It returns optional. No need to use `null`